tls/extensions/s2n_client_key_share.c

```diff
@@ -56,27 +57,92 @@ const s2n_extension_type s2n_client_key_share_extension = {
+        } else if ((key_shares >> i) & 1) { /* If bit other than lsb is set, generate keyshare for the corresponding curve */
+            GUARD(s2n_ecdhe_parameters_send(ecc_evp_params, out));
+        } else { /* If bit not set, skip keyshare generation for the corresponding curve */
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->iana_id));
+            GUARD(s2n_stuffer_write_uint16(out, ecc_evp_params->negotiated_curve->share_size));
+            GUARD(s2n_stuffer_skip_write(out, ecc_evp_params->negotiated_curve->share_size));
+        }
+    }
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_send_hrr_keyshare(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
     const struct s2n_ecc_named_curve *named_curve = NULL;
     struct s2n_ecc_evp_params *ecc_evp_params = NULL;
 
-    for (uint32_t i = 0; i < ecc_pref->count; i++) {
-        ecc_evp_params = &conn->secure.client_ecc_evp_params[i];
-        named_curve = ecc_pref->ecc_curves[i];
+    const struct s2n_ecc_preferences *ecc_pref = NULL;
+    GUARD(s2n_connection_get_ecc_preferences(conn, &ecc_pref));
+    notnull_check(ecc_pref);
 
-        ecc_evp_params->negotiated_curve = named_curve;
-        ecc_evp_params->evp_pkey = NULL;
-        GUARD(s2n_ecdhe_parameters_send(ecc_evp_params, out));
+    /* Our original key shares weren't succesful, so clear the old list of keyshares */
+    for (int i = 0; i < ecc_pref->count; i++) {
+        if (&conn->secure.client_ecc_evp_params[i] != NULL) {
+            GUARD(s2n_ecc_evp_params_free(&conn->secure.client_ecc_evp_params[i]));
+            conn->secure.client_ecc_evp_params[i].negotiated_curve = NULL;
+        }
     }
 
+    /* Generate the keyshare for the server negotiated curve */
+    ecc_evp_params = &conn->secure.client_ecc_evp_params[0];
+    named_curve = conn->secure.server_ecc_evp_params.negotiated_curve;
+    notnull_check(named_curve);
+
+    ecc_evp_params->negotiated_curve = named_curve;
+    ecc_evp_params->evp_pkey = NULL;
+    GUARD(s2n_ecdhe_parameters_send(ecc_evp_params, out));
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_ecdhe_supported_curves_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn);
+    notnull_check(conn->config);
+
+    /* From https://tools.ietf.org/html/rfc8446#section-4.1.2
+     * If a "key_share" extension was supplied in the HelloRetryRequest,
+     * replace the list of shares with a list containing a single
+     * KeyShareEntry from the indicated group.*/
+    if (s2n_server_hello_is_hrr(conn)) {
+        GUARD(s2n_send_hrr_keyshare(conn, out));
+        return S2N_SUCCESS;
+    }
+
+    if (!conn->preferred_key_shares) {
+        /* Default behavior is to generate keyshares for all curves.
+        * The bitmap to generate keyshares for all curve is 111111110 (254),
+        * i.e. all bit values set except lsb which is RESERVED for empty keyshares */
+        conn->preferred_key_shares = 254;
```

Comment:
Will changing this value have any side effects? Should we copy it first so we don't affect the original value?

Reply:
This value is set only for the default case (conn->preferred_key_shares = 0), i.e when the customer donot explicitly request keyshares and we are required to generate keyshares for all the curves in the preferences list.